### PR TITLE
[Open-Sora-Plan v1.3.0]: inference and training

### PR DESCRIPTION
**Pull request by wtomin**
- [x] vae inference & training under pynative mode;
- [x] dit inference & training under pynative mode;
- [x] dynamic resolution dataloader;
- [x] prompt refiner inference.

### CODE REVIEW

File: examples/opensora_pku/README.md [hunk 1]
```diff
@@ -498,62 +492,27 @@ We also support to run validation during training. This is supported by editing
...
-+  --lr_scheduler="polynomial_decay" \
-+  --lr_decay_steps=1000000 \
-```
-The edits will set the polynomial_decay LR scheduler, and decay the start LR to the end LR in 1000000 steps. You can adjust `lr_decay_steps` based on your `max_train_steps`. See other options of LR scheduler in `mindone/trainers/lr_schedule.py`.
+See `train_t2v_stage2.sh` under `scripts/text_condition/mult-devices/` for detailed usage.
 
 #### Performance
 
-The training performance are tested on ascend 910* with mindspore 2.3.1 graph mode. The results are as follows.
+We evaluated the training performance on Ascend NPUs. All experiments are running in PYNATIVE mode. The results are as follows.
```
Comment (SamitHuang) [inline examples/opensora_pku/README.md:505]: describe MS version.

File: examples/opensora_pku/README.md [hunk 2]
```diff
@@ -498,62 +492,27 @@ We also support to run validation during training. This is supported by editing
...
 
-| model name      | cards       |  stage     |batch size   | num frames| resolution  | graph compile | parallelism | recompute | sink | jit level| s/step | img/s |
-|:----------------|:----------- |:----------|:---------:|:-----:|:----------:|:----------:|:----------:|:----------:|:----------:|:----------:|-------------------:|:----------:|
-| OpenSoraT2V-ROPE-L-122 |  8   | 2 | 8  |           1 | 640x480     |  3mins     |         zero2                     | ON | ON | O0 |    2.35      | 27.23 |
-| OpenSoraT2V-ROPE-L-122 |  8   | 3 |  1  |          29 | 640x480    |    6mins    |       zero2                      |  ON | ON | O0 |     3.68     | 63.04 |
-| OpenSoraT2V-ROPE-L-122 |  8   | 4 |  1  |          29 |1280x720   |   10mins    |       zero2 + SP(sp_size=8)      |  OFF | ON | O0 |    4.32     | 6.71 |
-| OpenSoraT2V-ROPE-L-122 |  8   | 5 |  1  |          93 | 1280x720   |   15mins    |       zero2 + SP(sp_size=8)      |  ON | ON | O0 |    24.40     | 3.81  |
+| model name      | cards       |  stage     | batch size (global)   | video size  | Paramllelism |recompute |data sink | jit level| step time | train imgs/s |
```
Comment (SamitHuang) [inline examples/opensora_pku/README.md:507]: Explain what `train imgs/s` is and how it's computed.

File: examples/opensora_pku/README.md (hunk 1 above)
Comment (wtomin) [reply examples/opensora_pku/README.md]: Fixed.

File: examples/opensora_pku/README.md (hunk 2 above)
Comment (wtomin) [reply examples/opensora_pku/README.md:507]: Fixed.

File: examples/opensora_pku/opensora/utils/ms_utils.py [hunk 3]
```diff
@@ -170,3 +171,22 @@ def init_env(
     )
     initialize_sequence_parallel_state(sp_size)
     return rank_id, device_num
+
+
+@ms.jit_class
+class no_grad(_no_grad):
```
Comment (SamitHuang) [inline examples/opensora_pku/opensora/utils/ms_utils.py:177]: why limit it to pynative mode

File: examples/opensora_pku/opensora/npu_config.py [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+        return rank_id, device_num
+
+    def get_total_cores(self):
+        try:
+            total_cores = os.sysconf("SC_NPROCESSORS_ONLN")
+        except (AttributeError, ValueError):
+            total_cores = os.cpu_count()
+        return total_cores
+
+    def bind_thread_to_cpu(self):
```
Comment (SamitHuang) [inline examples/opensora_pku/opensora/npu_config.py:105]: same here

File: examples/opensora_pku/opensora/npu_config.py [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+        pid = os.getpid()
+        command = f"taskset -cp {cpu_cores_range} {pid}"
+
+        subprocess.run(command, shell=True, check=True)
+        return f"Binding Cores: {self.rank} {pid}  {cpu_cores_range}"
+
+    def get_attention_mask(self, attention_mask, repeat_num):
+        if self.on_npu and attention_mask is not None:
+            if npu_config.enable_FA:
+                attention_mask = attention_mask.to(ms.float16)
```
Comment (SamitHuang) [inline examples/opensora_pku/opensora/npu_config.py:125]: why fix to `ms.float16`

File: examples/opensora_pku/opensora/utils/ms_utils.py (hunk 3 above)
Comment (wtomin) [reply examples/opensora_pku/opensora/utils/ms_utils.py:177]: This class `no_grad` is copied from other projects, like movigen or opensora_hpcai. I guess it's because graph mode does not need to manually set no_grad.

File: examples/opensora_pku/opensora/npu_config.py (hunk 4 above)
Comment (wtomin) [reply examples/opensora_pku/opensora/npu_config.py]: Fixed

File: examples/opensora_pku/opensora/npu_config.py (hunk 5 above)
Comment (wtomin) [reply examples/opensora_pku/opensora/npu_config.py:98]: The `attention_mask` input to `get_attention_mask`  function is fp32. I think whether casting it to fp16 or bool, the purpose is to make the mask tensor size smaller.

I notice that in [the same function](https://github.com/PKU-YuanGroup/Open-Sora-Plan/blob/main/opensora/npu_config.py#L163C1-L168C30) in torch repo, they cast it to bool.
```python
    def get_attention_mask(self, attention_mask, repeat_num):
        if self.on_npu and attention_mask is not None:
            if npu_config.enable_FA:
                attention_mask = attention_mask.to(torch.bool)
            attention_mask = attention_mask.repeat_interleave(repeat_num, dim=-2)
        return attention_mask
```

I remember that calling `repeat_interleave` on ms.bool_ tensor is very slow, so I cast it to fp16 here.

When calling `ms_flash_attention`, the attention_mask will be further casted to `self.fa_mask_dtype`, which is usually `ms.uint8`.

**Review by SamitHuang — APPROVED** [review]

**Review by CaitinZhao — APPROVED** [review]
